### ISSUE TITLE
Event propagation doesn't comply to spec

### ISSUE THREAD

**Issue by jojastahl** [open62541/open62541#2474]
As far as I understand the current codebase, when an event is triggered, it is added to all "HasComponent" parents of the source node. This doesn't comply with the specification, IMHO. The specs define a HasNotifier / HasEventSource hierarchy for event propagation, if my understanding of OPC UA Part 3, 7.17 and 7.18 is correct. In addition, 7.17 says that the Server singleton object should report all events generated anywhere, regardsless of any HasEventSource references.

Is my point of view correct, or am I missing something?

**Comment by jpfr** [comment]
@Pro @aDogCalledSpot 
Can you confirm?

Can we make a meta-issue that collects things that need to be done for Events to get rid of the "experimental" status?

**Comment by ccvca** [comment]
The current implementation is the behaviour, which I would expect after diskussion this with several people about this topic.

![grafik](https://user-images.githubusercontent.com/3939648/53104741-37a87500-3530-11e9-85d0-fd0a477ebc73.png)
Speaks about that an Object ("Area") would deliver all Events from a subset, but it also says that this is just an example.

From 5.4:
![grafik](https://user-images.githubusercontent.com/3939648/53105217-fd8ba300-3530-11e9-8917-8994e70eab53.png)
So a mechanism to propagate events to parent nodes.

**Comment by jojastahl** [comment]
I agree that OPC UA Part 3 is a little soft on how to handle event propagation. However I feel that the existence of HasNotifier/HasEventSource and the samples given for these references in Part 3, aswell as Part 9 (Alarms and Conditions) points heavily in the direction to respect these. That's my personal feeling.

I looked at how others are doing: I compared with the Softing OPC UA Toolkit using this model:

![grafik](https://user-images.githubusercontent.com/15796520/53179908-eadaa200-35f4-11e9-960d-2b3ac3225087.png)

Result:

1. Raising an event on Alarms:
   - Softing: The event appears on Alarms and Server Object, 1x each.
   - Open62541: The event appears on Alarms, OPC-Test-System, DeviceSet and Server Object, 1x each
2. Raising an event on Conditions:
   - Softing: The event appears on Conditions, OPC-Test-System and Server Object, 1x each
   - Open62541: The event appears 1x on Conditions, and 2x on each of OPC-Test-System, DeviceSet and Server Object

Conclusion:
- Softing reports each event on the event source, and always on the server, too. Furthermore they are reported on objects in the HasNotifier/HasEventSource hierarchy.
- Open62541 reports on any parent which is enabled for events, and on the server. If there is a HasNotifier/HasEventSource reference, it is even reported twice.
- Open62541 reports an event raised on Conditions 2x on DeviceSet, even if there is no HasNotifier/HasEventSource link.

**Comment by jpfr** [comment]
Was this solved with this commit last year?
https://github.com/open62541/open62541/commit/e6baba149f4f66d3ddb40baacb91354e2d812509#diff-dd7eb0316f0b90c03ec8a6123b429054

**Comment by teccheck** [comment]
As there's no further reply, I'll close this issue. Feel free to reopen if needed
